### ISSUE TITLE
what is self_attn.head_out

### ISSUE THREAD

**Issue by tjlujianqiao** [likenneth/honest_llama#7]
Dear, I notice that you need to extract head attention in utils.py (function: get_llama_activations_bau) using the Tracedict method

However, there is no attribute named "model.layers.{i}.self_attn.head_out" in the llama model (actually you can only get q,k,v proj, and o proj). 

From my understanding, I guess the activations shall be the output of <softmax(q^T*k)v> right?, i.e., head output before projecting back.

Bests,

**Comment by likenneth** [comment]
Hi there, we modified the original llama modelling file to have this [head_out](https://github.com/likenneth/honest_llama/blob/master/llama/modeling_llama.py#L198]). It does not change model architecture, only marking a location where features are extracted, modified and put back.  
You can read the code for more detail.

**Comment by Sanster** [comment]
[baukit](https://github.com/davidbau/baukit/tree/main/baukit) can retain module's input, we can modify the trace code as following:

```python
# change from self_attn.head_out to self_attn.o_proj
HEADS = [f"model.layers.{i}.self_attn.o_proj" for i in range(model.config.num_hidden_layers)]

with TraceDict(model, HEADS+MLPS, retain_input=True) as ret:
    output = model(prompt, output_hidden_states = True)
    ...
    head_wise_hidden_states = [ret[head].input.squeeze().detach().cpu() for head in HEADS]
```

**Comment by likenneth** [comment]
@Sanster Thanks for letting me know!
